pkg/sysregistriesv2/alias.go

```diff
@@ -0,0 +1,112 @@
+package sysregistriesv2
+
+import (
+	"os"
+	"path/filepath"
+
+	"github.com/BurntSushi/toml"
+	"github.com/containers/image/v5/types"
+	"github.com/containers/storage/pkg/homedir"
+	"github.com/containers/storage/pkg/lockfile"
+)
+
+// userAliasesFile is the user-specific registries.conf file to store aliases.
+var userAliasesFile = filepath.FromSlash(".config/containers/aliases.conf")
+
+// userAliasesLockFile is the lockfile for the user-specific registries.conf
+// file to store aliases.
+var userAliasesLockFile = filepath.FromSlash(".local/share/containers/aliases.conf.lock")
+
+// ResolveAlias performs an alias resolution of the specified name.  The
+// user-specific aliases.conf has precedence over aliases in `conf`.
+func (conf *V2RegistriesConf) ResolveAlias(ctx *types.SystemContext, name string) (string, bool, error) {
+	confPath, lock, err := aliasesConfPathAndLock(ctx)
+	if err != nil {
+		return "", false, err
+	}
+
+	// Acquire the lock as a reader to allow for multiple routines in the
+	// same process space to read in simultaneously.
+	lock.RLock()
+	defer lock.Unlock()
+
+	aliasConf, err := loadAliasesConf(confPath)
+	if err != nil {
+		return "", false, err
+	}
+
+	// Note: aliases.conf has precedence.
+	value, resolved := aliasConf.Aliases[name]
+	if !resolved {
+		value, resolved = conf.Aliases[name]
+	}
+	return value, resolved, nil
+}
+
+// AddAliases adds the specified name-value pair as a new alias to the
+// user-specific aliases.conf.  It may override an existing alias for `name`.
+func (conf *V2RegistriesConf) AddAlias(ctx *types.SystemContext, name string, value string) error {
+	confPath, lock, err := aliasesConfPathAndLock(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Acquire the lock as a writer to prevent data corruption.
+	lock.Lock()
+	defer lock.Unlock()
+
+	// Load the aliases.conf, add the specified name-value pair, and write
+	// it back to the file.
+	aliasConf, err := loadAliasesConf(confPath)
+	if err != nil {
+		return err
+	}
+
+	aliasConf.Aliases[name] = value
+
+	f, err := os.OpenFile(confPath, os.O_RDWR|os.O_CREATE, 0600)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	encoder := toml.NewEncoder(f)
+	return encoder.Encode(aliasConf)
+}
+
+func loadAliasesConf(confPath string) (*aliasesConf, error) {
+	conf := aliasesConf{}
+
+	_, err := toml.DecodeFile(confPath, &conf)
+	if os.IsNotExist(err) {
+		// It's okay to not have an aliases.conf (yet), so just ignore
+		// it for now.
+		err = nil
+	}
+
+	if err == nil && conf.Aliases == nil {
+		conf.Aliases = make(map[string]string)
+	}
+
+	return &conf, err
+}
+
+func aliasesConfPathAndLock(ctx *types.SystemContext) (string, lockfile.Locker, error) {
+	aliasesConfPath := ""
+	aliasesConfLockPath := ""
+
+	if ctx != nil && len(ctx.UserAliasesConfPath) > 0 {
+		aliasesConfPath = ctx.UserAliasesConfPath
+	} else {
+		aliasesConfPath = filepath.Join(homedir.Get(), userAliasesFile)
+	}
+
+	if ctx != nil && len(ctx.UserAliasesConfLockPath) > 0 {
```

Comment:
I don’t think this should be configurable; the same configuration should always use the same lock file. Just hard-code a suffix.